jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpTransportOverHTTP2.java

```diff
@@ -350,119 +351,363 @@ public void abort(Throwable failure)
+     *
+     * @see State
+     */
     private class TransportCallback implements Callback
     {
-        private State state = State.IDLE;
-        private Callback callback;
-        private Throwable failure;
-        private boolean commit;
+        private State _state = State.IDLE;
+        private Callback _callback;
+        private boolean _commit;
+        private Throwable _failure;
 
-        public boolean start(Callback callback, boolean commit)
+        private void reset()
         {
-            State state;
+            assert Thread.holdsLock(this);
+            _state = State.IDLE;
+            _callback = null;
+            _commit = false;
+            _failure = null;
+        }
+
+        private void send(Callback callback, boolean commit, Consumer<Callback> sendFrame)
+        {
+            Throwable failure = sending(callback, commit);
+            if (failure == null)
+            {
+                sendFrame.accept(this);
+                pending();
+            }
+            else
+            {
+                callback.failed(failure);
+            }
+        }
+
+        private Throwable sending(Callback callback, boolean commit)
+        {
+            synchronized (this)
+            {
+                switch (_state)
+                {
+                    case IDLE:
+                    {
+                        _state = State.SENDING;
+                        _callback = callback;
+                        _commit = commit;
+                        return null;
+                    }
+                    case FAILED:
+                    {
+                        return _failure;
+                    }
+                    default:
+                    {
+                        return new IllegalStateException("Invalid transport state: " + _state);
+                    }
+                }
+            }
+        }
+
+        private void pending()
+        {
+            Callback callback;
+            boolean commit;
             Throwable failure;
             synchronized (this)
             {
-                state = this.state;
-                failure = this.failure;
-                if (state == State.IDLE)
+                switch (_state)
                 {
-                    this.state = State.WRITING;
-                    this.callback = callback;
-                    this.commit = commit;
-                    return true;
+                    case SENDING:
+                    {
+                        // The send has not completed the callback yet,
+                        // wait for succeeded() or failed() to be called.
+                        _state = State.PENDING;
+                        return;
+                    }
+                    case SUCCEEDING:
+                    {
+                        // The send already completed successfully, but the
+                        // call to succeeded() was delayed, so call it now.
+                        callback = _callback;
```

Comment:
I think we should null `_callback` whenever we take it.

Reply:
It is in `reset()` few lines below.